Public/Src/Utilities/Storage/InputChange/InputChangeList.cs

```diff
@@ -0,0 +1,195 @@
+                }
+            }
+            catch (IOException ioException)
+            {
+                Logger.Log.ExceptionOnCreatingInputChangeList(loggingContext, path, ioException.ToString());
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="InputChangeList"/> from a stream reader.
+        /// </summary>
+        public static InputChangeList CreateFromStream(LoggingContext loggingContext, TextReader reader, string optionalPath = null)
+        {
+            Contract.Requires(loggingContext != null);
+            Contract.Requires(reader != null);
+
+            InputChangeList inputChangeList = new InputChangeList();
+
+            try
+            {   
+                int lineNo = 0;
+
+                while (reader.Peek() >= 0)
+                {
+                    string inputLine = reader.ReadLine();
+                    ++lineNo;
+
+                    if (string.IsNullOrEmpty(inputLine))
+                    {
+                        continue;
+                    }
+
+                    if (!TryParseInput(loggingContext, inputLine, optionalPath ?? string.Empty, lineNo, out var changePathInfo))
+                    {
+                        return null;
+                    }
+
+                    inputChangeList.m_changedPaths.Add(changePathInfo);
+                }
+            }
+            catch (IOException ioException)
+            {
+                Logger.Log.ExceptionOnCreatingInputChangeList(loggingContext, optionalPath ?? string.Empty, ioException.ToString());
+                return null;
+            }
+
+            return inputChangeList;
+        }
+
+        /// <summary>
+        /// Tries to parse input line.
+        /// </summary>
+        /// <remarks>
+        /// Input line must be of the form:
+        ///   full path
+        /// or
+        ///   full path|comma separated <see cref="PathChanges"/>
+        /// The former assumes that changes are <see cref="PathChanges.DataOrMetadataChanged"/>.
+        /// </remarks>
+        private static bool TryParseInput(LoggingContext loggingContext, string input, string filePath, int lineNo, out ChangedPathInfo changedPathInfo)
+        {
+            Contract.Requires(loggingContext != null);
+            Contract.Requires(!string.IsNullOrEmpty(input));
+
+            changedPathInfo = default;
+            string[] splitInput = input.Split(s_inputSeparator);
+
+            if (splitInput.Length > 2)
+            {
+                Logger.Log.InvalidFormatOfInputChange(loggingContext, input, filePath, lineNo);
+                return false;
+            }
+
+            string changedPath = splitInput[0].Trim();
+            string changesStr = null;
+
+            if (splitInput.Length == 2)
+            {
+                changesStr = splitInput[1].Trim();
+            }
+
+            if (!Path.IsPathRooted(changedPath))

```

Comment:
This can fail for illegal file characters...

Path.IsPathRooted("\a") throws....